typing/includecore.ml

```diff
@@ -374,9 +374,15 @@ let compare_records_with_representation ~loc env params1 params2 n
   =
   match compare_records ~loc env params1 params2 n labels1 labels2 with
   | None when rep1 <> rep2 ->
-      let pos = if rep2 = Record_float then Second else First in
-      Some (Unboxed_float_representation pos)
-  | err -> err
+      Some (match rep1, rep2 with
+            | _, Record_unboxed _ -> Unboxed_representation Second
+            | Record_unboxed _, _ -> Unboxed_representation First
+            | _, Record_float ->
+               Record_mismatch (Unboxed_float_representation Second)
+            | _, _ ->
+               Record_mismatch (Unboxed_float_representation First))
```

Comment:
If I understand correctly, this change fixes a bug in the `Regular, Unboxed` case for example. Is this correct?

The code (old or new) is making assumptions about the fact that the `Record_inlined` and `Record_extension` constructors will not show up here. Are those assumptions correct? Are they documented somewhere?

I wonder if we could write a non-fragile pattern-match instead of first check `rep1 <> rep2` and then coming up with a pithy list of relevant cases.

Reply:
I wrote the pattern match, it is nicer. There's still an assumption that Record_inlined and Record_extension don't mix, but now it's an explicit assert false rather than silently converting errors to weird messages about floats.

(For the record, I believe the assumption here is true because it's impossible to name an inlined record in the source language, and the compiler keeps the autogenerated inline record types separate)